# ImageProcessing1
## Load Image
In order to load image from a file, you need to import library. The library in need is called "scikit-image".<br>
"scikit-image"(or skimage) is a collection of many features relating to scientific image processing, and many of the features are saved to sub directories of skimage. Input/output library is save to skimage.io, so you have to import like the following code.
```python
import skimage.io as ski
```

In [2]:
import skimage.io as ski

To read image, use ```skimage.io.imread(filename)```. ```filename``` is a string of filename. It can be both relative path or absolute path.<br>
```python
img=ski.imread("testimage.tif")
```

In [3]:
img=ski.imread("testimage.tif")

Print the image using print().<br>
Image is processed as a 2-dementional numpy array in python, thus printing image will show the array of numbers.
```python
print(img)
```

In [4]:
print(img)

[[ 64  66  68 ... 122 124 126]
 [ 66  68  70 ... 124 126 128]
 [ 68  70  72 ... 126 128 130]
 ...
 [122 124 126 ... 180 182 184]
 [124 126 128 ... 182 184 186]
 [126 128 130 ... 184 186 188]]


Check the type of ```img```. It must be <class 'numpy.ndarray'>.
```python
print(type(img))
```

In [5]:
print(type(img))

<class 'numpy.ndarray'>


Check the ```img.dtype```. This image is saved as 8-bit, so the dtype must be uint8.
```python
print(img.dtype)
```

In [6]:
print(img.dtype)

uint8


## Save Image
You can save the image by ```skimage.io.imsave(filename)```. ```filename``` is a string of filename. It can be both relative path or absolute path.<br>
```python
ski.imsave("result.tif", img)
```

In [7]:
ski.imsave("result.tif", img)

## Simple Image Processing(With ROI)
You understand that ```img``` is numpy-array. So, you can now do the image processing just like to modify numpy-array.<br>
First, let's create a ROI and fill that ROI with zero.
```python
ROI=img[14:18,14:18]
print(ROI)
```

In [8]:
ROI=img[14:18,14:18]
print(ROI)

[[120 122 124 126]
 [122 124 126 128]
 [124 126 128 130]
 [126 128 130 132]]


Next, we fill the ROI with 0.<br>
You can do this by writing ```ROI=numpy.zeros((4,4))```, but there is another way to do this. It is multiplying 0. This can convert all numbers in ROI to 0.<br>
Note that ```ROI``` is created as a **shallow-copy** of the original ```img```, so the calculation to ```ROI``` will change ```img``` as well.
```python
ROI*=0
```

In [9]:
ROI*=0

Save the image with the name of "result.tif".(This will overwrite and save the image).
```python
ski.imsave("result.tif", img)
```

In [10]:
ski.imsave("result.tif", img)

Open "result.tif" with your image-viewer. You can see that the image has a black rectangle in the center.

## Changing the Contranst of Image
Reload testimage with
```python
img=ski.imread("testimage.tif")
```

In [11]:
img=ski.imread("testimage.tif")

Numpy has some statistical functions, and amoung them are ```.min()``` and ```.max()```. These functions can find minimal, maximal number in the numpy-array, respectivly.
```python
minPixel=img.min()
maxPixel=img.max()
print(minPixel)
print(maxPixel)
```

In [12]:
minPixel=img.min()
maxPixel=img.max()
print(minPixel)
print(maxPixel)

64
188


The darkest pixel in testimage.tif is 64, and the lightest is 188.
To create the high-contrast image from original image, we should convert 64 to 0, and 188 to 255. We do this by......
>Subtract 188 from each pixel. ```newimg=img-minPixel```<br>
>Multiply 255/(maxPixel-minPixel) to each pixel. ```scale=(255/(maxPixel-minPixel))``` and ```newimg=newimg*scale```<br>

In [13]:
scale=(255/(maxPixel-minPixel))
newimg=img-minPixel
newimg=newimg*scale

OK, you can now have the high-contrast image!<br>
All done? Actually, no.<br>
check the dtype of ```newimage```.
```python
print(newimg.dtype)
```

In [14]:
print(newimg.dtype)

float64


The type is now float64! In this format, you cannot save it! (It must be uint8)<br>
But why? What makes the image-type to change? The committer is ```newimg=newimg*scale```. In this calculation, you are multipling uint8 array and float. This will automatically convert the type of numpy-array to a suitable type for calculation.<br>
<br>
So, what shall we do? We must convert the type by using ```.astype(code)```. This will change the datatype of array.
```python
newimg=newimg.astype("u1")
print(newimg.dtype)
```

In [15]:
newimg=newimg.astype("u1")
print(newimg.dtype)

uint8


Ok, the type is now "uint8". Now you get the high-contrast image. Save it by running
```python
ski.imsave("result.tif", newimg)
```

In [16]:
ski.imsave("result.tif", newimg)

Open "result.tif" with your image-viewer. You can see that the image becomes high-contrasted.